### PR TITLE
add terminators

### PR DESCRIPTION
**Pull request by im-scared**
This change introduces Terminators as described in [this comment](https://github.com/JulienPalard/Pipe/issues/74#issuecomment-1436481100) that I made on issue #74.

I discovered that the `>` operator can be used for terminators, as it has lower precedence than `|` so it will properly terminate the pipeline. I chose to use a separate operator for terminators in order to reduce misunderstandings as discussed on #74. I also placed terminators in their own package to further differentiate them, so users will have more opportunities to understand that they are used differently.

I've also implemented a whole bunch of common terminators I see used in other languages, especially java (my main language).

I also just remembered the README needs to be updated to explain terminators, too. I'm happy to make that addition if the concept is approved.

P.S.
I'm not the most skilled python user, so I may have done the module handling and `__init__` file stuff wrong ;_;. But if so I'm happy to do whatever's needed to fix it!

### CODE REVIEW

File: pipe/terminator.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+
+@Terminator
+def first(iterable):
+    return next(iter(iterable))
+
+
+@Terminator
+def last(iterable):
+    return deque(iterable, maxlen=1).pop()
```
Comment (im-scared) [inline pipe/terminator.py:120]: These will raise exceptions if the iterable is empty. I didn't love that, but it does seem to be the common pattern in python? In java streams, these functions would return `Optional` values, but I couldn't find anything like that in python >_<. Maybe it's fine like this, but I wanted to call out the concern.

**Comment by JulienPalard** [comment]
I don't know if I would accept it but your PR gave me yet another idea (that we can't even implement easily), something around the lines of just adding this to `Pipe`:

```python
    def __lt__(self, function):
        return function(self)
```

So that `range(10) | take(3) > list` would just call `list` on the `Pipe`. This way we would not have to implement any terminator: they're just functions!

But but but... in my library, `range(10) | take(3)` is not a Pipe instance, it's (naturally ?) a native `generator` instance, so no operator overloading here.

File: pipe/terminator.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+def _identity(x):
+    return x
+
+
+@Terminator
+def as_dict(iterable, key_func=_identity, value_func=_identity):
+    if key_func is _identity and value_func is _identity:
+        return dict(iterable)
+
+    return dict(iterable | pipe.map(lambda x: (key_func(x), value_func(x))))
```
Comment (JulienPalard) [inline pipe/terminator.py:66]: I think the as_dict should it a Pipe of pairs, instead of duplicating the value, like:

    [(1, 2), (3, 4), (5, 6)] > as_dict == {1: 2, 3: 4, 5: 6}

This change your test from:

```python
    data = ["banana", "grapefruit", "guava"]
    expected = {"BANANA": 6, "GRAPEFRUIT": 10, "GUAVA": 5}
    actual = data > term.as_dict(
        key_func=lambda e: e.upper(),
        value_func=lambda e: len(e),
    )
    assert isinstance(actual, dict)
    assert actual == expected
```

to:

```python
    data = ["banana", "grapefruit", "guava"]
    expected = {"BANANA": 6, "GRAPEFRUIT": 10, "GUAVA": 5}
    actual = data | map(lambda fruit: (fruit.upper(), len(fruit)) > term.as_dict
    assert isinstance(actual, dict)
    assert actual == expected
```

Which I find more natural. It may completly eliminate the need of key_func and value_func?

File: pipe/terminator.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+
+@Terminator
+def as_list(iterable):
+    return list(iterable)
+
+
+@Terminator
+def as_set(iterable, key_func=_identity):
+    if key_func is _identity:
```
Comment (JulienPalard) [inline pipe/terminator.py:76]: Same here, I feel like the key_func can be dropped, as the same can be expressed by using `map`:

    [1, 2, 3] > as_set(key_func=lambda x: x ** 2) == [1, 2, 3] | map(lambda x: x ** 2) | as_set

**Comment by JulienPalard** [comment]
It feel strange that terminators can be chained too, I mean (not tried):

    >>> range(10) | map(str) > join("-") > first
    "a"

(OK it is a bit far-fetched).

I've not taken a lot of time to think about it yet, but I think I would be happy to see your `terminator` module packaged as its own package.

So people would be able to just `pip install pipe` or to `pip install pipe terminators` depending on what they want.

If you do so, I'll gladly add a link to your project from my README, obviously.

And as it makes things very explicit (the install step, the import step, and for (I hope) the majority of project, a proper import, so an explicit `| term.`). And as terminators can also be chained. And as it would be your project. I would have nothing against your library using `|` and not `>`. It could literraly depend on `pipe` and import `Pipe` to use it.

File: pipe/terminator.py (hunk 2 above)
Comment (im-scared) [reply pipe/terminator.py:66]: Oh that is entirely in line with how I intended `as_dict` and `as_set` to work :D. I am copying the idea from java streams, where you can provide a key or value function if you want, or if you already have a stream of pairs then you can just call `as_dict` with no arguments. (It's technically not that clean in java, but basically like that)

So the intention was that `key_func` and `value_func` exist as a convenience if writing your terminator like that would be more clear vs using `map`, but you absolutely can just call `as_dict` with no arguments if you have a pipe of pairs or would prefer to use `map`.

**Comment by im-scared** [comment]
> It feel strange that terminators can be chained too

Yeah, that's technically true. People would just have to not do that. Additionally, if someone tried to do something big brain like the following, it wouldn't work how they expect because of operator precedence (this is a ridiculous example but illustrates the point):
```python
range(10) | map(str) > as_list | is_even
```

In fact, I already ran into a version of this in the tests, because the following wouldn't work since `==` evaluated before `>`:
```python
assert range(10) | is_even > last == 8
```

Pipe already has some edge cases like this, I think? Though they aren't as likely to happen. But I figured it could be addressed with proper documentation. However, it does make me like your suggestion:

> ...I think I would be happy to see your `terminator` module packaged as its own package.
> 
> So people would be able to just `pip install pipe` or to `pip install pipe terminators` depending on what they want.
> ...
> ...I would have nothing against your library using `|` and not `>`

Having a separate module could definitely make it a lot more clear to people how to use terminators, so they don't trip as much! And getting to still use `|` would solve the operator precedence issue :D. So if you'd prefer that I'd be happy to do it! It'll be my first time publishing anything public like a module so a great learning experience, too :D.

One thought about the module, though. What do you feel about calling it `pipe-terminators`, since it'll be pretty specific to pipes, even if it could be used without using any pipe functions.

**Comment by JulienPalard** [comment]
> Pipe already has some edge cases like this, I think?

Yes, worse case probably being playing with set unions and pipes together ;) Exposing an operator overload is always highly risky.

> So if you'd prefer that I'd be happy to do it! It'll be my first time publishing anything public like a module so a great learning experience, too :D.

I'm a a big fan of using small (or not small) projects like this as a pretext to learn :) I'm using https://git.afpy.org/AFPy/python-docs-fr/ a lot as a pretext to teach how to contribute to free software to people around me.

> One thought about the module, though. What do you feel about calling it `pipe-terminators`, since it'll be pretty specific to pipes, even if it could be used without using any pipe functions.

Works for me!

I'll be offline most of the time for the following week (holidays here) but don't hesitate to ping me (you'll find my personal email address in the commits) if you have questions. The way `pipe` is packaged looks up to date to me (no old `setup.py`, no old `setup.cfg`, no unmaintained dependency in `pyproject.toml`), I'd say you can take inspiration from it if needed. But the interesting point is that if you prefer flit or poetry or whatever, as it's you project, you can use it without me interfering ;)

**Comment by im-scared** [comment]
Sounds like a plan! I'll put together a more complete module, then, and let you know once I've released it! :D
